src/App.tsx

```diff
@@ -8,7 +8,69 @@ import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
 
+import { Todo } from './types/Todo';
+import { FilterType } from './types/FilterType';
+import { getTodos } from './api';
+
+const filterTodos = (todos: Todo[], inputQuery: string, filterType: FilterType) => {
+  let currentTodos = [...todos];
+
+  if (inputQuery) {
+    const query = inputQuery.trim().toLowerCase();
+
+    currentTodos = currentTodos.filter(todo => todo.title.toLowerCase().includes(query));
+  }
+
+  switch (filterType) {
+    case FilterType.ACTIVE:
+      currentTodos = currentTodos.filter(todo => !todo.completed);
+      break;
+
+    case FilterType.COMPLETED:
+      currentTodos = currentTodos.filter(todo => todo.completed);
+      break;
+
+    default:
+      throw new Error(`Unrecognized filter type: ${filterType}`);
+  }
+
+  return currentTodos;
+};
+
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [isError, setIsError] = useState(false);
+  const [isDataLoading, setIsDataLoading] = useState(true);
+  const [query, setQuery] = useState<string>('');
+  const [filterType, setFilterType] = useState(FilterType.ALL);
+
+  const visibleTodos = filterTodos(todos, query, filterType);
+
+  const fetchTodos = async () => {
+    try {
+      const data = await getTodos();
+
+      setTodos(data);
+      setIsDataLoading(false);
+      setIsDataLoading(false);
```

Comment:
why do you do it twice here? And what about this state if an error occurs, should it stay true?